### PR TITLE
On trait bound mismatch, detect multiple crate versions in dep tree

### PR DESCRIPTION
**Pull request by estebank**
When encountering an E0277, if the type and the trait both come from a crate with the same name but different crate number, we explain that there are multiple crate versions in the dependency tree.

If there's a type that fulfills the bound, and it has the same name as the passed in type and has the same crate name, we explain that the same type in two different versions of the same crate *are different*.

```
error[E0277]: the trait bound `Type: dependency::Trait` is not satisfied
 --> src/main.rs:4:18
  |
4 |     do_something(Type);
  |     ------------ ^^^^ the trait `dependency::Trait` is not implemented for `Type`
  |     |
  |     required by a bound introduced by this call
  |
help: you have multiple different versions of crate `dependency` in your dependency graph
 --> src/main.rs:1:5
  |
1 | use bar::do_something;
  |     ^^^ one version of crate `dependency` is used here, as a dependency of crate `bar`
2 | use dependency::Type;
  |     ^^^^^^^^^^ one version of crate `dependency` is used here, as a direct dependency of the current crate
note: two types coming from two different versions of the same crate are different types even if they look the same
 --> /home/gh-estebank/crate_versions/baz-2/src/lib.rs:1:1
  |
1 | pub struct Type;
  | ^^^^^^^^^^^^^^^ this type doesn't implement the required trait
  |
 ::: /home/gh-estebank/crate_versions/baz/src/lib.rs:1:1
  |
1 | pub struct Type;
  | ^^^^^^^^^^^^^^^ this type implements the required trait
2 | pub trait Trait {}
  | --------------- this is the required trait
note: required by a bound in `bar::do_something`
 --> /home/gh-estebank/crate_versions/baz/src/lib.rs:4:24
  |
4 | pub fn do_something<X: Trait>(_: X) {}
  |                        ^^^^^ required by this bound in `do_something`
```

Address #22750.

### CODE REVIEW

**Review by fee1-dead — APPROVED** [review]
Looks great! It would be nice if we could add a test for this though. Maybe [run-make](https://rustc-dev-guide.rust-lang.org/tests/compiletest.html#run-make-tests)?

**Comment by weiznich** [comment]
This new error message looks great :heart: 

I think it would even be more helpful if it would point the user to use `cargo tree` as debugging tool for finding the sources of the duplicated crates in their dependency tree. From my experience from helping users with such issues they are often not aware that they have several versions of the same crate in their dependency tree. Sometimes the even question that this might be the issue. This is almost always resolved by pointing them to `cargo tree -p $crate -i`, although that command will likely not work out of the box as it requires specifying the version of `$crate` if it appears in different versions in your dependency tree. By default `cargo tree` will print a error message that contains the different versions that occur in your tree, so the user has at least a starting point from there. 

Something else I noticed with this error message is that it only describes what's the problem, but it does not really mention how to resolve it. Maybe include something that essentially says: Try to restructure your dependency tree in such way that this uses the same version of that crate here?

**Comment by estebank** [comment]
@weiznich [added a note about `cargo tree`, but it gets somewhat lost in the noise](https://github.com/rust-lang/rust/pull/124944/files#diff-4070e8fafbce87b6f888eddd3efd65d37450b862616a417be2b8d3f6bdce6424) :-/

**Comment by estebank** [comment]
@fee1-dead added test

**Comment by fee1-dead** [comment]
Looks great! Thanks for working on this!

**Comment by estebank** [comment]
Need to fix the tests, clearly they worked because of dirty environment only. I'll fix them.

**Comment by fee1-dead** [comment]
Do the tests work now? If they do r=me

**Comment by estebank** [comment]
@fee1-dead I haven't been able to replicate the CI failure in my dev env, which has stumped me.

**Comment by fee1-dead** [comment]
If you want, you could modify the try CI config and use [at]bors try to replicate the CI test environment

**Comment by estebank** [comment]
The issue was that the test was being run on the wasm platform, which doesn't produce the expected object files.

**Comment by fee1-dead** [comment]
Note that bors try r= will merge the PR without doing the full tests. Please don't do that again

@bors r-

**Comment by matthiaskrgr** [comment]
@bors r- 
https://github.com/rust-lang/rust/pull/127646#issuecomment-2225322251

**Comment by estebank** [comment]
> Note that bors try r= will merge the PR without doing the full tests. Please wait for try to finish before marking the PR approved next time

I wasn't familiar with that behavior. Thanks for letting me know.

----

I'll try one more time restricting this just to one platform, but I'm otherwise stumped on what the right way to setup this test would be while still getting the color output. A make test would be my next option.

File: tests/run-make/crate-loading/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+//@ only-linux
+//@ ignore-wasm32
```
Comment (jieyouxu) [inline tests/run-make/crate-loading/rmake.rs:2]: Question: does this only fail on wasm32, or does it also fail on wasm 64 targets? We could run `test-various` and `x86_64-gnu-llvm-17` try jobs to double check.

File: tests/run-make/crate-loading/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        "you have multiple different versions of crate `dependency` in your dependency graph",
+    );
+    assert_contains(
+        &stderr,
+        "two types coming from two different versions of the same crate are different types even \
+         if they look the same",
+    );
+    assert_contains(&stderr, "this type doesn't implement the required trait");
+    assert_contains(&stderr, "this type implements the required trait");
+    assert_contains(&stderr, "this is the required trait");
```
Comment (Oneirical) [inline tests/run-make/crate-loading/rmake.rs:29]: Suggestion: 

```rs
        .run_fail()
        .assert_stderr_contains("you have multiple different versions of crate `dependency` in your dependency graph")
        .assert_stderr_contains("two types coming from two different versions of the same crate are different types even \
         if they look the same")
        .assert_stderr_contains("this type doesn't implement the required trait")
        .assert_stderr_contains("this type implements the required trait")
        .assert_stderr_contains("this is the required trait");
```

**Comment by Kixunil** [comment]
This message is absolutely amazing! Thanks a ton @estebank for this work! :heart:

**Pull request merged**